File: src/main/java/no/fintlabs/validation/valueparsability/validators/DynamicStringParsabilityValidator.java
package no.fintlabs.validation.valueparsability.validators;

import no.fintlabs.model.configuration.entities.ValueMapping;
import no.fintlabs.validation.valueparsability.ValueParsabilityValidator;
import org.springframework.stereotype.Service;

import java.util.regex.Pattern;

@Service
public class DynamicStringParsabilityValidator implements ValueParsabilityValidator {

    private static final Pattern textPattern = Pattern.compile("(?:(?!\\$if\\{).)*");
    private static final Pattern instanceValueKeyPattern = Pattern.compile("(?:(?!\\$if\\{).)+");
    private static final Pattern ifReferencePattern = Pattern.compile("(?:\\$if\\{" + instanceValueKeyPattern + "})*");
    private static final Pattern dynamicStringPattern = Pattern.compile("^(?:" + textPattern + "|" + ifReferencePattern + ")*$");

    @Override
    public ValueMapping.Type getTypeToValidate() {
        return ValueMapping.Type.DYNAMIC_STRING;
    }

    @Override
    public boolean isValid(String value) {
        return dynamicStringPattern.matcher(value).matches();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/valueparsability/validators/DynamicStringParsabilityValidator.kt
package no.novari.flyt.configuration.validation.valueparsability.validators

import no.novari.flyt.configuration.model.configuration.entities.ValueMapping
import no.novari.flyt.configuration.validation.valueparsability.ValueParsabilityValidator
import org.springframework.stereotype.Service
import java.util.regex.Pattern

@Service
class DynamicStringParsabilityValidator : ValueParsabilityValidator {
    private val textPattern = Pattern.compile("(?:(?!${Pattern.quote(IF_REFERENCE_PREFIX)}).)*")
    private val instanceValueKeyPattern = Pattern.compile("(?:(?!${Pattern.quote(IF_REFERENCE_PREFIX)}).)+")
    private val ifReferencePattern =
        Pattern.compile(
            "(?:${Pattern.quote(IF_REFERENCE_PREFIX)}$instanceValueKeyPattern\\})*",
        )
    private val dynamicStringPattern = Pattern.compile("^(?:$textPattern|$ifReferencePattern)*$")

    override fun getTypeToValidate(): ValueMapping.Type = ValueMapping.Type.DYNAMIC_STRING

    override fun isValid(value: String): Boolean = dynamicStringPattern.matcher(value).matches()

    private companion object {
        const val IF_REFERENCE_PREFIX = "\$if{"
    }
}
